### PR TITLE
🛠Improve xrefmap.json de-serialization

### PR DESCRIPTION
**Pull request by live1206**
- decompress the file during restore while `contentEncoding` is `gzip`
- improve the performance of de-serialize `xrefmap.json`

### CODE REVIEW

File: src/docfx/restore/RestoreFile.cs [hunk 1]
```diff
@@ -100,7 +101,18 @@ public static string GetRestoreEtagPath(string url)
                 using (var stream = await response.EnsureSuccessStatusCode().Content.ReadAsStreamAsync())
                 using (var file = new FileStream(tempFile, FileMode.Create, FileAccess.Write, FileShare.None))
                 {
-                    await stream.CopyToAsync(file);
+                    string contentEncoding = response.Content.Headers.ContentEncoding.ToString();
```
Comment (yufeih) [inline src/docfx/restore/RestoreFile.cs]: https://stackoverflow.com/a/27327208

File: docs/specs/xref.yml [hunk 2]
```diff
@@ -21,21 +21,14 @@ outputs:
...
-    { 
-      "references":[{
-          "uid": "System.String",
-          "name": "String",
-          "fullName": "System.String",
-          "href": "https://docs.microsoft.com/en-us/dotnet/api/system.string",
-          "nameWithType": "System.String"
-      }]
-    }
+    { "references":[{"uid": "System.String", "name": "String", "fullName": "System.String", "href": "https://docs.microsoft.com/en-us/dotnet/api/system.string", "nameWithType": "System.String"}]}
```
Comment (OsmondJiang) [inline docs/specs/xref.yml]: no need change?

File: src/docfx/build/xref/XrefMap.cs [hunk 3]
```diff
@@ -152,25 +156,71 @@ private bool TryResolveFromExternal(string uid, out XrefSpec spec)
...
+                        if (json.TokenType == JsonToken.StartObject)
+                        {
+                            startPosition = json.LinePosition;
+                        }
+                        else if (json.TokenType == JsonToken.EndObject)
+                        {
+                            endPosition = json.LinePosition;
+                            if (uid != null)
+                            {
+                                map[uid] = new Lazy<XrefSpec>(() => JsonUtility.Deserialize<XrefSpec>(content.Substring(startPosition - 1, endPosition - startPosition + 1)));
```
Comment (OsmondJiang) [inline src/docfx/build/xref/XrefMap.cs]: will the lazy function take too much spaces since every one has a content string copy?

Comment (live1206) [reply src/docfx/build/xref/XrefMap.cs]: not really, every spec only has their `UID` definition
```json
{"uid":"Microsoft.Office.Interop.Outlook","name":"Microsoft.Office.Interop.Outlook","href":"https://docs.microsoft.com/dotnet/api/microsoft.office.interop.outlook","commentId":"N:Microsoft.Office.Interop.Outlook","fullName":"Microsoft.Office.Interop.Outlook","nameWithType":"Microsoft.Office.Interop.Outlook"}
```

File: docs/specs/xref.yml [hunk 4]
```diff
@@ -684,3 +684,4 @@ outputs:
     ["info","at-uid-not-found","Cannot find uid 'f' using xref '<xref href=\"f\" data-raw-source=\"@mentionf\">'","docs/a.md",5,12]
     ["info","at-uid-not-found","Cannot find uid 'e' using xref '<p>\n  <xref href=\"e\" data-raw-source=\"@mentione\" />\n  some text\n  <xref href=\"f\" data-raw-source=\"mentionf\" />\n</p>'","docs/a.md",6,18]
     ["warning","uid-not-found","Cannot find uid 'f' using xref '<p>\n  <xref href=\"e\" data-raw-source=\"@mentione\" />\n  some text\n  <xref href=\"f\" data-raw-source=\"mentionf\" />\n</p>'","docs/a.md",6,78]
+---
```
Comment (OsmondJiang) [inline docs/specs/xref.yml]: remove

File: src/docfx/lib/HttpClientUtility.cs [hunk 5]
```diff
@@ -11,7 +11,10 @@ namespace Microsoft.Docs.Build
 {
     internal static class HttpClientUtility
     {
-        private static readonly HttpClient s_httpClient = new HttpClient();
+        private static readonly HttpClient s_httpClient = new HttpClient(new HttpClientHandler()
+        {
+            AutomaticDecompression = DecompressionMethods.GZip | DecompressionMethods.Deflate,
```
Comment (OsmondJiang) [inline src/docfx/lib/HttpClientUtility.cs]: will this make it as our default http client handler?

File: src/docfx/build/xref/XrefMap.cs [hunk 6]
```diff
@@ -152,24 +156,28 @@ private bool TryResolveFromExternal(string uid, out XrefSpec spec)
 
         public static XrefMap Create(Context context, Docset docset)
         {
-            Dictionary<string, XrefSpec> map = new Dictionary<string, XrefSpec>();
-            foreach (var url in docset.Config.Xref)
+            // TODO: not considering same uid with multiple specs, it will be Dictionary<string, List<T>>
+            // https://github.com/dotnet/corefx/issues/12067
```
Comment (yufeih) [inline src/docfx/build/xref/XrefMap.cs]: 👍 want to replace all `ConcurrentBags`, `ConcurrentDictionary` with non concurrent ones and locks?

File: src/docfx/build/xref/XrefMap.cs [hunk 7]
```diff
@@ -152,24 +156,28 @@ private bool TryResolveFromExternal(string uid, out XrefSpec spec)
...
+            ParallelUtility.ForEach(docset.Config.Xref, url =>
             {
                 var (_, content, _) = RestoreMap.GetRestoredFileContent(docset, url);
                 XrefMapModel xrefMap = new XrefMapModel();
                 if (url?.Value.EndsWith(".yml", StringComparison.OrdinalIgnoreCase) != false)
                 {
                     xrefMap = YamlUtility.Deserialize<XrefMapModel>(content);
+                    foreach (var spec in xrefMap.References)
+                    {
+                        map[spec.Uid] = new Lazy<XrefSpec>(() => spec);
```
Comment (yufeih) [inline src/docfx/build/xref/XrefMap.cs]: lock

File: src/docfx/build/xref/XrefMap.cs (hunk 6 above)
Comment (live1206) [reply src/docfx/build/xref/XrefMap.cs]: That depends on the scenario while reading and adding, `ConcurrentDictionary` is better. But sure, we can take a look at the other usages

**Pull request merged**